fix-27998 - [Windows] ScrollView is not scrolling to the bottom if in grid with *,auto Width

BrayanKhosravian opened the pull request:
fixes: https://github.com/dotnet/maui/issues/27998

**Problems:**
|---cell0---|
|cell1|cell2|

**Problem1:**
All row and column defs are either set to auto or *.
Cell2 has an absolute width set to 250.
After initialization of the cells this absolute value can be considered for the row and column defs.
**Solution1:**
We consider the abolsute values set for the children which have to be layouted in the column and row defs.

**Problem2:**
Whenever we measured we have to reconsider that the row and column defs should be adjusted.
**Solution2:**
Whenver we measured (which is currently after `FirstMeasurePass()` and `SecoondMeasurePass()`) we adjust the row and column defs and consider possible overflows.

BrayanKhosravian committed:
-fix-27998: added UI Tests

BrayanKhosravian committed:
-fix-27998: added and changed unit tests

BrayanKhosravian committed:
-fix-27998: fixed issue

BrayanKhosravian commented:
When reviewing my own code then I can think of removeing following calls:
`definitions.Sum(d => d.Size);`
`definitions.Where(d => d.IsAuto || d.IsStar).OrderByDescending(d => d.Size).FirstOrDefault()`
Which would mean that we could also remove the introduced namespace:
`using System.Linq;`

For that to happen I would have to wrap `Definitions[]` and store following values:
`public bool MaxDefinitionSize {get; private set;}`
`public Definition BiggestDefinition {get; private set;}`

The wrapper would have to act as sort of proxy.
This mean we would have to limit access to be able to change values in a Definition object.
Because the wrapper would have to do this, then update the changes to the `MaxDefinitionSize` and `BiggestDefinition`.

At this point we could call `AdjustDefinitions()` after every `MeasureCell()` call instead of calling it once after `FirstMeasurePass()` and `SecondMeasurePass()`.

BrayanKhosravian commented:
This is only my attempt of solveing it.
It is fine for me in case someone copies or cherry-picks the uitest and unittest commits and fixes it differently.

jsuarezruiz commented:
/azp run

jsuarezruiz commented on src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue27998.cs:
```diff
@@ -0,0 +1,39 @@
+		[Category(UITestCategories.Cells)]
+		[Category(UITestCategories.ScrollView)]
+		[Category(UITestCategories.Border)]
+		public void GridAutosStarsScollToEndDisplaysLastItem()
```
The test is failing on Android, iOS and Mac.

```
Assert.That(lastItem.IsDisplayed().Equals(false), Is.True)
Expected: True
But was:  False
```

Could you take a look?

BrayanKhosravian replied:
Sorry for not responding for a long while.
I have been drowning in uni activities.
But was also stuck as the compilation failed for ios when paired with a macincloud device.
I will open a discussion regarding the issue I faced.

But my testcase succeds now for android.

Open tasks for myself:
- check if testcase succeeds on ios and mac
- investigate the other failing tests

BrayanKhosravian committed:
fix-27998: test for android at least

BrayanKhosravian committed:
fix-27998 cleanup

BrayanKhosravian commented:
What an awesome review and tooling.
Have been off here for a while.
I will have a look at this.

BrayanKhosravian committed:
implemented review

BrayanKhosravian committed:
fix-27998: review

kubaflo reviewed (changes requested):
Could you please check the ai's suggestions?

BrayanKhosravian commented:
I have been hesitating a bit because I reinvestigated the issue deeper and I am unsure if the code I want to merge actually only treats a symptom rather then fixing the root cause.
But I want to share my findings.

The last commit broke the "ScrollView Mixed Feature Tests" which had Grid in horizontal options.
That worked with my original "fix" "[-fix-27998: fixed issue](https://github.com/dotnet/maui/pull/28258/changes/05fa3ec68177507b14d005478d406be82545dff9)".
Also until this commit the solution worked: [implemented review](https://github.com/dotnet/maui/pull/28258/changes/dfb6fe0eeb1b3b682dd73e6a7bfdbf2d46845baa)
Basically my last commit broke it.

But I think this is actually not required and handles a completely different issue, where the scrollbar is not visible for windows.
```cs
// some _childrenToLayout could have set Height/Width abosultely set, while our columns/rows are on auto or *
// in this case we set set the column/row size to the child's size
foreach (var cell in _cells)
{
	var view = _childrenToLayOut[cell.ViewIndex];
	if (view.Height > 0)
	{
		var row = _rows[cell.Row];
		if (row.IsAuto)
		{
			row.Size = view.Height;
		}
	}
	if (view.Width > 0)
	{
		var column = _columns[cell.Column];
		if (column.IsAuto)
		{
			column.Size = view.Width;
		}
	}
}
```

As I looked into the ScrollView Mixed Feature Tests, I noticed other scrolling issues as well.
Then checked out the main branch and noticed that horizontal scrolling does not work (at least on windows, did not check on android),
Because of this and the Review responses I came to the point where I think that the actual issue(es) is/are located somewhere else.

Therefore I added debug logs before and after measure passes, insinde arragne semantics, but also in the code behind of the Controls and logged their SizeChanged and their Handler values.
Will apply the logs at the end.
Based on this logs I think that the problem is in the GridLayoutManager and not in any handler.
Android gets also a wrong measure value but ignores it or clips it to its screen constraints.

The AI review mentioned hot path and performance.
I found a solution which works better through overflow detextion and requires only a call in the 2nd pass but still foreach cell. This is not optimal but overflows shouldnt occur often. After an overflow was detected the affected definitions and cels are corrected.

But because of this I came to the point and asked why do this values have to be corrected?
And currently I am trying to find a solution which already uses more correct values for the 2nd measure pass.
But this involves a bigger refactoring to make the code clearer.

Currently I have few pararell branches open regarding the same issue.

Regarding the overflow detection I do something like this:
(In SecondMeasurePass the only thing I added is `TryCorrectOverflow`)
```cs
void SecondMeasurePass()
{
	foreach (var cell in _cells)
	{
		if (!cell.NeedsSecondPass)
		{
			continue;
		}

		double width = 0;
		double height = 0;

		if (double.IsInfinity(cell.MeasureHeight))
		{
			height = double.PositiveInfinity;
		}
		else
		{
			for (int n = cell.Row; n < cell.Row + cell.RowSpan; n++)
			{

				height += _rows.GetSize(n);
			}
			height += _rowSpacing * (cell.RowSpan > 0 ? cell.RowSpan - 1 : 0);
		}

		if (double.IsInfinity(cell.MeasureWidth))
		{
			width = double.PositiveInfinity;
		}
		else
		{
			for (int n = cell.Column; n < cell.Column + cell.ColumnSpan; n++)
			{
				width += _columns.GetSize(n);
			}
			width += _columnSpacing * (cell.ColumnSpan > 0 ? cell.ColumnSpan - 1 : 0);
		}

		if (width == 0 || height == 0)
		{
			continue;
		}

		var measure = MeasureCell(cell, width, height);

		if (cell.IsColumnSpanStar && cell.ColumnSpan > 1)
		{
			TrackSpan(new Span(cell.Column, cell.ColumnSpan, true, measure.Width));
		}
		else if (cell.ColumnSpan == 1)
		{
			if (TreatCellWidthAsAuto(cell))
			{
				_columns.Update(cell.Column, measure.Width);
			}
			else
			{

			}
		}

		if (cell.IsRowSpanStar && cell.RowSpan > 1)
		{
			TrackSpan(new Span(cell.Row, cell.RowSpan, false, measure.Height));
		}
		else if (cell.RowSpan == 1)
		{
			if (TreatCellHeightAsAuto(cell))
			{
				_rows.Update(cell.Row, measure.Height);
			}
			else
			{

			}
		}
		TryCorrectOverflow();

	}
}

void TryCorrectOverflow()
{
	if (!_rows.DetectOverflow(_gridHeightConstraint))
		return;

	if (!_rows.TryCorrectOverflow(_gridHeightConstraint, out int affectedRowIndex, out double? correctedSize))
		return;

	if (correctedSize is null)
		return;

	for (int i = 0; i < _cells.Length; i++)
	{
		var cell = _cells[i];
		if (cell.Row != affectedRowIndex)
			continue;

		cell.MeasureHeight = correctedSize.Value;
	}
	
}
```

But to make this work I created a new class DefinitionStructure which is used for rows and columns.
It keeps trackof the sizes changed and the caller can call for OverflowDetection and Correction.

```cs
sealed class DefinitionStructure
{

//...
private double _changedSize;
//...

public void Update(int index, double size)
{
	var definition = _definitions[index];
	ChangeSize(definition, size);
}

public void ChangeSize(int index, double size) => 
	ChangeSize(_definitions[index], size);

void ChangeSize(Definition def, double size)
{
	_changedSize += def.SetSize(size);
}


public bool DetectOverflow(double constraint) => 
	_changedSize > constraint;

public bool TryCorrectOverflow(double constraint, out int affectedIndex, out double? correctedSize)
{
	affectedIndex = -1;
	correctedSize = null;
	var overflow = _changedSize - constraint;
	if (overflow <= 0)
		return false;

	Definition? biggestDefinition = null;
	for (var i = 0; i < _definitions.Length; i++)
	{
		Definition? definition = _definitions[i];
		if (!definition.IsAuto && !definition.IsStar)
			continue;
		if (biggestDefinition is null || definition.Size > biggestDefinition.Size)
		{
			biggestDefinition = definition;
			affectedIndex = i;
		}
	}
	if (biggestDefinition is null)
		return false;

	correctedSize = Math.Max(0, biggestDefinition.Size - overflow);
	ChangeSize(biggestDefinition, correctedSize.Value);


	return true;
}
//...
```

I wanted to share my thoughts and analysis.
Regarding the additional Debug Output (This is from the main branch):
GS - GridStructure
GLM - GridLayoutManager
27998 - prefix are logs in the code behind throught subscribing .SizeChanged event handler

Windows Debug WriteLines
```
LayoutPanel.ArrangeOverride finalSize=1008,728
LayoutPanel.ArrangeOverride finalSize=1008,728
LayoutPanel.ArrangeOverride actual={Width=1008 Height=728}
LayoutPanel.ArrangeOverride actual={Width=1008 Height=728}
LayoutPanel.ArrangeOverride finalSize=1008,728
LayoutPanel.ArrangeOverride finalSize=1008,728
LayoutPanel.ArrangeOverride actual={Width=1008 Height=728}
LayoutPanel.ArrangeOverride actual={Width=1008 Height=728}
LayoutPanel.ArrangeOverride finalSize=1008,692
LayoutPanel.ArrangeOverride finalSize=1008,692
LayoutPanel.ArrangeOverride actual={Width=1008 Height=692}
LayoutPanel.ArrangeOverride actual={Width=1008 Height=692}
Start: GS: ctor before Init
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: ctor before Init
Start: GS: ctor after Init
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: ctor after Init
Start: GS: After first measrue pass
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: After first measrue pass
Start: GS: star columns resolved
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: star columns resolved
Start: GS: star rows resolved
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=728
End: GS: star rows resolved
GS: before Measure Cell: row=0 column=0 width=1008 height=∞
GS: after Measure Cell: row=0 column=0 sizeResult={Width=326 Height=96}
GS: before Measure Cell: row=1 column=0 width=1008 height=728
GS: after Measure Cell: row=1 column=0 sizeResult={Width=1008 Height=728}
GS: before Measure Cell: row=1 column=1 width=∞ height=728
GS: after Measure Cell: row=1 column=1 sizeResult={Width=50 Height=2}
Start: GS: after second measure pass
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96
Row: 1Type=Star Measure=728
End: GS: after second measure pass
Start: GS: after resolve spans
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96
Row: 1Type=Star Measure=728
End: GS: after resolve spans
Start: GS: after MinimizeStarsForMeasurement
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96
Row: 1Type=Star Measure=728
End: GS: after MinimizeStarsForMeasurement
GLM.MeasureING: widthConstrain=1008 heightConstraint=728
GLM.MeasurED: measuredWidth=1058 measuredHeight=824
LayoutPanel.ArrangeOverride finalSize=1008,728
27998: Grid: 1008 x 728
LayoutPanel.ArrangeOverride finalSize=1058,824
GLM.ArrangeING: bounds={X=0 Y=0 Width=1058 Height=824}
Start: GS: before PrepareForArrange
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96
Row: 1Type=Star Measure=728
End: GS: before PrepareForArrange
Start: GS: inside PrepareForArrange after MinimizeStar Rows and Columns
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96
Row: 1Type=Star Measure=728
End: GS: inside PrepareForArrange after MinimizeStar Rows and Columns
Start: GS: after PrepareForArrange after expand star columns and rows
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96
Row: 1Type=Star Measure=728
End: GS: after PrepareForArrange after expand star columns and rows
GLM.cellBounds: height=96 width=1058 location={X=0 Y=0}
LayoutPanel.ArrangeOverride finalSize=1058,96
LayoutPanel.ArrangeOverride actual={Width=1058 Height=96}
GLM.arrangedView: height=NaN width=NaN transitionX=0 transitionY=0 anchorX=0,5 anchorX=0,5
GLM.cellBounds: height=728 width=1008 location={X=0 Y=96}
27998: ScrollView: 1008 x 728
GLM.arrangedView: height=NaN width=NaN transitionX=0 transitionY=0 anchorX=0,5 anchorX=0,5
GLM.cellBounds: height=728 width=50 location={X=1008 Y=96}
27998: Border: 50 x 728
GLM.arrangedView: height=NaN width=50 transitionX=0 transitionY=0 anchorX=0,5 anchorX=0,5
GLM.ArrangED: actualSize={Width=1058 Height=824}
GLM.ArrangED: adjustedActualSize={Width=1058 Height=824}
LayoutPanel.ArrangeOverride actual={Width=1058 Height=824}
LayoutPanel.ArrangeOverride actual={Width=1008 Height=728}
Start: GS: ctor before Init
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: ctor before Init
Start: GS: ctor after Init
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: ctor after Init
Start: GS: After first measrue pass
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: After first measrue pass
Start: GS: star columns resolved
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: star columns resolved
Start: GS: star rows resolved
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=728
End: GS: star rows resolved
GS: before Measure Cell: row=0 column=0 width=1008 height=∞
GS: after Measure Cell: row=0 column=0 sizeResult={Width=351 Height=96}
GS: before Measure Cell: row=1 column=0 width=1008 height=728
GS: after Measure Cell: row=1 column=0 sizeResult={Width=1008 Height=728}
GS: before Measure Cell: row=1 column=1 width=∞ height=728
GS: after Measure Cell: row=1 column=1 sizeResult={Width=50 Height=2}
Start: GS: after second measure pass
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96
Row: 1Type=Star Measure=728
End: GS: after second measure pass
Start: GS: after resolve spans
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96
Row: 1Type=Star Measure=728
End: GS: after resolve spans
Start: GS: after MinimizeStarsForMeasurement
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96
Row: 1Type=Star Measure=728
End: GS: after MinimizeStarsForMeasurement
GLM.MeasureING: widthConstrain=1008 heightConstraint=728
GLM.MeasurED: measuredWidth=1058 measuredHeight=824
LayoutPanel.ArrangeOverride finalSize=1008,728
LayoutPanel.ArrangeOverride finalSize=1058,824
GLM.ArrangeING: bounds={X=0 Y=0 Width=1058 Height=824}
Start: GS: before PrepareForArrange
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96
Row: 1Type=Star Measure=728
End: GS: before PrepareForArrange
Start: GS: inside PrepareForArrange after MinimizeStar Rows and Columns
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96
Row: 1Type=Star Measure=728
End: GS: inside PrepareForArrange after MinimizeStar Rows and Columns
Start: GS: after PrepareForArrange after expand star columns and rows
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96
Row: 1Type=Star Measure=728
End: GS: after PrepareForArrange after expand star columns and rows
GLM.cellBounds: height=96 width=1058 location={X=0 Y=0}
LayoutPanel.ArrangeOverride finalSize=1058,96
LayoutPanel.ArrangeOverride actual={Width=1058 Height=96}
GLM.arrangedView: height=NaN width=NaN transitionX=0 transitionY=0 anchorX=0,5 anchorX=0,5
GLM.cellBounds: height=728 width=1008 location={X=0 Y=96}
GLM.arrangedView: height=NaN width=NaN transitionX=0 transitionY=0 anchorX=0,5 anchorX=0,5
GLM.cellBounds: height=728 width=50 location={X=1008 Y=96}
GLM.arrangedView: height=NaN width=50 transitionX=0 transitionY=0 anchorX=0,5 anchorX=0,5
GLM.ArrangED: actualSize={Width=1058 Height=824}
GLM.ArrangED: adjustedActualSize={Width=1058 Height=824}
LayoutPanel.ArrangeOverride actual={Width=1058 Height=824}
LayoutPanel.ArrangeOverride actual={Width=1008 Height=728}
LayoutPanel.ArrangeOverride finalSize=1008,728
LayoutPanel.ArrangeOverride finalSize=1058,824
GLM.ArrangeING: bounds={X=0 Y=0 Width=1058 Height=824}
Start: GS: before PrepareForArrange
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96
Row: 1Type=Star Measure=728
End: GS: before PrepareForArrange
Start: GS: inside PrepareForArrange after MinimizeStar Rows and Columns
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96
Row: 1Type=Star Measure=728
End: GS: inside PrepareForArrange after MinimizeStar Rows and Columns
Start: GS: after PrepareForArrange after expand star columns and rows
Column: 0Type=Star Measure=1008
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96
Row: 1Type=Star Measure=728
End: GS: after PrepareForArrange after expand star columns and rows
GLM.cellBounds: height=96 width=1058 location={X=0 Y=0}
GLM.arrangedView: height=NaN width=NaN transitionX=0 transitionY=0 anchorX=0,5 anchorX=0,5
GLM.cellBounds: height=728 width=1008 location={X=0 Y=96}
LayoutPanel.ArrangeOverride finalSize=1008,1026
27998: Content: 1008 x 1026
LayoutPanel.ArrangeOverride finalSize=1008,1026
LayoutPanel.ArrangeOverride actual={Width=1008 Height=1026}
LayoutPanel.ArrangeOverride actual={Width=1008 Height=1026}
GLM.arrangedView: height=NaN width=NaN transitionX=0 transitionY=0 anchorX=0,5 anchorX=0,5
GLM.cellBounds: height=728 width=50 location={X=1008 Y=96}
LayoutPanel.ArrangeOverride finalSize=50,728
LayoutPanel.ArrangeOverride actual={Width=50 Height=728}
GLM.arrangedView: height=NaN width=50 transitionX=0 transitionY=0 anchorX=0,5 anchorX=0,5
GLM.ArrangED: actualSize={Width=1058 Height=824}
GLM.ArrangED: adjustedActualSize={Width=1058 Height=824}
LayoutPanel.ArrangeOverride actual={Width=1058 Height=824}
LayoutPanel.ArrangeOverride actual={Width=1008 Height=728}
LayoutPanel.ArrangeOverride finalSize=50,728
LayoutPanel.ArrangeOverride actual={Width=50 Height=728}
27998: Page: 1008 x 728
27998: Grid Bounds: {X=0 Y=0 Width=1008 Height=728}
27998: ScrollView Bounds: {X=0 Y=96 Width=1008 Height=728}
27998: Border Bounds: {X=1008 Y=96 Width=50 Height=728}
Native Grid: 1058 x 824
Native Scroll: 1008 x 728
27998: === GRID ===
27998: Bounds      : {X=0 Y=0 Width=1008 Height=728}
27998: DesiredSize : {Width=1008 Height=728}
27998: === CHILDREN ===
27998: VerticalStackLayout       Bounds={X=0 Y=0 Width=1058 Height=96} Desired={Width=351 Height=96}
27998: ScrollView                Bounds={X=0 Y=96 Width=1008 Height=728} Desired={Width=1008 Height=728}
27998: Border                    Bounds={X=1008 Y=96 Width=50 Height=728} Desired={Width=50 Height=2}
LayoutPanel.ArrangeOverride finalSize=1008,1026
LayoutPanel.ArrangeOverride actual={Width=1008 Height=1026}
```


For Android:
```
ich hab noch nichts verändert aber nur mal zum vergleich, das sind die logs für android:
Start: GS: ctor before Init
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: ctor before Init
Start: GS: ctor after Init
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: ctor after Init
Start: GS: After first measrue pass
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: After first measrue pass
Start: GS: star columns resolved
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: star columns resolved
Start: GS: star rows resolved
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=845.7142857142857
End: GS: star rows resolved
GS: before Measure Cell: row=0 column=0 width=411.42857142857144 height=∞
GS: after Measure Cell: row=0 column=0 sizeResult={Width=324.85714285714283 Height=96.57142857142857}
GS: before Measure Cell: row=1 column=0 width=411.42857142857144 height=845.7142857142857
GS: after Measure Cell: row=1 column=0 sizeResult={Width=411.42857142857144 Height=845.7142857142857}
GS: before Measure Cell: row=1 column=1 width=∞ height=845.7142857142857
GS: after Measure Cell: row=1 column=1 sizeResult={Width=50 Height=2}
Start: GS: after second measure pass
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96.57142857142857
Row: 1Type=Star Measure=845.7142857142857
End: GS: after second measure pass
Start: GS: after resolve spans
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96.57142857142857
Row: 1Type=Star Measure=845.7142857142857
End: GS: after resolve spans
Start: GS: after MinimizeStarsForMeasurement
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96.57142857142857
Row: 1Type=Star Measure=845.7142857142857
End: GS: after MinimizeStarsForMeasurement
GLM.MeasureING: widthConstrain=411.42857142857144 heightConstraint=845.7142857142857
GLM.MeasurED: measuredWidth=461.42857142857144 measuredHeight=942.2857142857142
Start: GS: ctor before Init
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: ctor before Init
Start: GS: ctor after Init
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: ctor after Init
Start: GS: After first measrue pass
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: After first measrue pass
Start: GS: star columns resolved
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: star columns resolved
Start: GS: star rows resolved
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=782.8571428571428
End: GS: star rows resolved
GS: before Measure Cell: row=0 column=0 width=411.42857142857144 height=∞
GS: after Measure Cell: row=0 column=0 sizeResult={Width=324.85714285714283 Height=96.57142857142857}
GS: before Measure Cell: row=1 column=0 width=411.42857142857144 height=782.8571428571428
GS: after Measure Cell: row=1 column=0 sizeResult={Width=411.42857142857144 Height=782.8571428571429}
GS: before Measure Cell: row=1 column=1 width=∞ height=782.8571428571428
GS: after Measure Cell: row=1 column=1 sizeResult={Width=50 Height=2}
Start: GS: after second measure pass
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96.57142857142857
Row: 1Type=Star Measure=782.8571428571428
End: GS: after second measure pass
Start: GS: after resolve spans
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96.57142857142857
Row: 1Type=Star Measure=782.8571428571428
End: GS: after resolve spans
Start: GS: after MinimizeStarsForMeasurement
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96.57142857142857
Row: 1Type=Star Measure=782.8571428571428
End: GS: after MinimizeStarsForMeasurement
GLM.MeasureING: widthConstrain=411.42857142857144 heightConstraint=782.8571428571428
GLM.MeasurED: measuredWidth=461.42857142857144 measuredHeight=879.4285714285713
27998: Grid: 411.42857142857144 x 845.7142857142857
GLM.ArrangeING: bounds={X=0 Y=38.857142857142854 Width=411.42857142857144 Height=782.8571428571428}
Start: GS: before PrepareForArrange
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96.57142857142857
Row: 1Type=Star Measure=782.8571428571428
End: GS: before PrepareForArrange
Start: GS: inside PrepareForArrange after MinimizeStar Rows and Columns
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96.57142857142857
Row: 1Type=Star Measure=782.8571428571428
End: GS: inside PrepareForArrange after MinimizeStar Rows and Columns
Start: GS: after PrepareForArrange after expand star columns and rows
Column: 0Type=Star Measure=361.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=96.57142857142857
Row: 1Type=Star Measure=686.2857142857142
End: GS: after PrepareForArrange after expand star columns and rows
GLM.cellBounds: height=96.57142857142857 width=411.42857142857144 location={X=0 Y=38.857142857142854}
GLM.arrangedView: height=NaN width=NaN transitionX=0 transitionY=0 anchorX=0.5 anchorX=0.5
GLM.cellBounds: height=686.2857142857142 width=361.42857142857144 location={X=0 Y=135.42857142857142}
27998: ScrollView: 361.42857142857144 x 686.2857142857142
27998: Content: 411.42857142857144 x 1010.5714285714286
GLM.arrangedView: height=NaN width=NaN transitionX=0 transitionY=0 anchorX=0.5 anchorX=0.5
GLM.cellBounds: height=686.2857142857142 width=50 location={X=361.42857142857144 Y=135.42857142857142}
27998: Border: 50 x 686.2857142857142
GLM.arrangedView: height=NaN width=50 transitionX=0 transitionY=0 anchorX=0.5 anchorX=0.5
GLM.ArrangED: actualSize={Width=411.42857142857144 Height=782.8571428571428}
GLM.ArrangED: adjustedActualSize={Width=411.42857142857144 Height=782.8571428571428}
Start: GS: ctor before Init
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: ctor before Init
Start: GS: ctor after Init
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: ctor after Init
Start: GS: After first measrue pass
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: After first measrue pass
Start: GS: star columns resolved
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: star columns resolved
Start: GS: star rows resolved
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=782.8571428571428
End: GS: star rows resolved
GS: before Measure Cell: row=0 column=0 width=411.42857142857144 height=∞
GS: after Measure Cell: row=0 column=0 sizeResult={Width=411.42857142857144 Height=181.14285714285714}
GS: before Measure Cell: row=1 column=0 width=411.42857142857144 height=782.8571428571428
GS: after Measure Cell: row=1 column=0 sizeResult={Width=411.42857142857144 Height=782.8571428571429}
GS: before Measure Cell: row=1 column=1 width=∞ height=782.8571428571428
GS: after Measure Cell: row=1 column=1 sizeResult={Width=50 Height=2}
Start: GS: after second measure pass
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=181.14285714285714
Row: 1Type=Star Measure=782.8571428571428
End: GS: after second measure pass
Start: GS: after resolve spans
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=181.14285714285714
Row: 1Type=Star Measure=782.8571428571428
End: GS: after resolve spans
Start: GS: after MinimizeStarsForMeasurement
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=181.14285714285714
Row: 1Type=Star Measure=782.8571428571428
End: GS: after MinimizeStarsForMeasurement
GLM.MeasureING: widthConstrain=411.42857142857144 heightConstraint=782.8571428571428
GLM.MeasurED: measuredWidth=461.42857142857144 measuredHeight=963.9999999999999
GLM.ArrangeING: bounds={X=0 Y=38.857142857142854 Width=411.42857142857144 Height=782.8571428571428}
Start: GS: before PrepareForArrange
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=181.14285714285714
Row: 1Type=Star Measure=782.8571428571428
End: GS: before PrepareForArrange
Start: GS: inside PrepareForArrange after MinimizeStar Rows and Columns
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=181.14285714285714
Row: 1Type=Star Measure=782.8571428571428
End: GS: inside PrepareForArrange after MinimizeStar Rows and Columns
Start: GS: after PrepareForArrange after expand star columns and rows
Column: 0Type=Star Measure=361.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=181.14285714285714
Row: 1Type=Star Measure=601.7142857142857
End: GS: after PrepareForArrange after expand star columns and rows
GLM.cellBounds: height=181.14285714285714 width=411.42857142857144 location={X=0 Y=38.857142857142854}
GLM.arrangedView: height=NaN width=NaN transitionX=0 transitionY=0 anchorX=0.5 anchorX=0.5
GLM.cellBounds: height=601.7142857142857 width=361.42857142857144 location={X=0 Y=220}
27998: ScrollView: 361.42857142857144 x 601.7142857142857
GLM.arrangedView: height=NaN width=NaN transitionX=0 transitionY=0 anchorX=0.5 anchorX=0.5
GLM.cellBounds: height=601.7142857142857 width=50 location={X=361.42857142857144 Y=220}
27998: Border: 50 x 601.7142857142857
GLM.arrangedView: height=NaN width=50 transitionX=0 transitionY=0 anchorX=0.5 anchorX=0.5
GLM.ArrangED: actualSize={Width=411.42857142857144 Height=782.8571428571428}
GLM.ArrangED: adjustedActualSize={Width=411.42857142857144 Height=782.8571428571428}
27998: Page: 411.42857142857144 x 845.7142857142857
27998: Grid Bounds: {X=0 Y=0 Width=411.42857142857144 Height=845.7142857142857}
27998: ScrollView Bounds: {X=0 Y=220 Width=361.42857142857144 Height=601.7142857142857}
27998: Border Bounds: {X=361.42857142857144 Y=220 Width=50 Height=601.7142857142857}
27998: === GRID ===
27998: Bounds      : {X=0 Y=0 Width=411.42857142857144 Height=845.7142857142857}
27998: DesiredSize : {Width=461.42857142857144 Height=1026.857142857143}
27998: === CHILDREN ===
27998: VerticalStackLayout       Bounds={X=0 Y=38.857142857142854 Width=411.42857142857144 Height=181.14285714285714} Desired={Width=411.42857142857144 Height=181.14285714285714}
27998: ScrollView                Bounds={X=0 Y=220 Width=361.42857142857144 Height=601.7142857142857} Desired={Width=411.42857142857144 Height=782.8571428571429}
27998: Border                    Bounds={X=361.42857142857144 Y=220 Width=50 Height=601.7142857142857} Desired={Width=50 Height=2}
Start: GS: ctor before Init
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: ctor before Init
Start: GS: ctor after Init
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: ctor after Init
Start: GS: After first measrue pass
Column: 0Type=Star Measure=0
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: After first measrue pass
Start: GS: star columns resolved
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=0
End: GS: star columns resolved
Start: GS: star rows resolved
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=0
Row: 0Type=Auto Measure=0
Row: 1Type=Star Measure=782.8571428571428
End: GS: star rows resolved
GS: before Measure Cell: row=0 column=0 width=411.42857142857144 height=∞
GS: after Measure Cell: row=0 column=0 sizeResult={Width=411.42857142857144 Height=181.14285714285714}
GS: before Measure Cell: row=1 column=0 width=411.42857142857144 height=782.8571428571428
GS: after Measure Cell: row=1 column=0 sizeResult={Width=411.42857142857144 Height=782.8571428571429}
GS: before Measure Cell: row=1 column=1 width=∞ height=782.8571428571428
GS: after Measure Cell: row=1 column=1 sizeResult={Width=50 Height=2}
Start: GS: after second measure pass
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=181.14285714285714
Row: 1Type=Star Measure=782.8571428571428
End: GS: after second measure pass
Start: GS: after resolve spans
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=181.14285714285714
Row: 1Type=Star Measure=782.8571428571428
End: GS: after resolve spans
Start: GS: after MinimizeStarsForMeasurement
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=181.14285714285714
Row: 1Type=Star Measure=782.8571428571428
End: GS: after MinimizeStarsForMeasurement
GLM.MeasureING: widthConstrain=411.42857142857144 heightConstraint=782.8571428571428
GLM.MeasurED: measuredWidth=461.42857142857144 measuredHeight=963.9999999999999
GLM.ArrangeING: bounds={X=0 Y=38.857142857142854 Width=411.42857142857144 Height=782.8571428571428}
Start: GS: before PrepareForArrange
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=181.14285714285714
Row: 1Type=Star Measure=782.8571428571428
End: GS: before PrepareForArrange
Start: GS: inside PrepareForArrange after MinimizeStar Rows and Columns
Column: 0Type=Star Measure=411.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=181.14285714285714
Row: 1Type=Star Measure=782.8571428571428
End: GS: inside PrepareForArrange after MinimizeStar Rows and Columns
Start: GS: after PrepareForArrange after expand star columns and rows
Column: 0Type=Star Measure=361.42857142857144
Column: 1Type=Auto Measure=50
Row: 0Type=Auto Measure=181.14285714285714
Row: 1Type=Star Measure=601.7142857142857
End: GS: after PrepareForArrange after expand star columns and rows
GLM.cellBounds: height=181.14285714285714 width=411.42857142857144 location={X=0 Y=38.857142857142854}
GLM.arrangedView: height=NaN width=NaN transitionX=0 transitionY=0 anchorX=0.5 anchorX=0.5
GLM.cellBounds: height=601.7142857142857 width=361.42857142857144 location={X=0 Y=220}
GLM.arrangedView: height=NaN width=NaN transitionX=0 transitionY=0 anchorX=0.5 anchorX=0.5
GLM.cellBounds: height=601.7142857142857 width=50 location={X=361.42857142857144 Y=220}
GLM.arrangedView: height=NaN width=50 transitionX=0 transitionY=0 anchorX=0.5 anchorX=0.5
GLM.ArrangED: actualSize={Width=411.42857142857144 Height=782.8571428571428}
GLM.ArrangED: adjustedActualSize={Width=411.42857142857144 Height=782.8571428571428}
The thread '.NET TP Worker' (29) has exited with code 0 (0x0).
The thread '.NET TP Worker' (26) has exited with code 0 (0x0).
The thread '.NET TP Worker' (24) has exited with code 0 (0x0).
The thread '.NET TP Worker' (19) has exited with code 0 (0x0).
The thread '.NET TP Worker' (23) has exited with code 0 (0x0).
The thread '.NET TP Worker' (28) has exited with code 0 (0x0).
The thread '.NET TP Worker' (20) has exited with code 0 (0x0).
The thread '.NET TP Worker' (30) has exited with code 0 (0x0).
The thread '.NET TP Worker' (32) has exited with code 0 (0x0).
The thread '.NET TP Worker' (22) has exited with code 0 (0x0).
The thread '.NET TP Worker' (25) has exited with code 0 (0x0).
The thread '.NET TP Worker' (27) has exited with code 0 (0x0).
The thread '.NET TP Worker' (21) has exited with code 0 (0x0).
The thread '.NET TP Worker' (31) has exited with code 0 (0x0).


```

kubaflo reviewed (changes requested):
It doesn't looks like a fix, but rather a workaround - could you please check the ai's suggestions? Maybe it is actually worth to close this PR?
